ethcore/src/engines/authority_round/mod.rs

```diff
@@ -125,54 +148,87 @@ impl From<ethjson::spec::AuthorityRoundParams> for AuthorityRoundParams {
 	}
 }
 
-// Helper for managing the step.
+/// Helper for managing the step.
 #[derive(Debug)]
 struct Step {
 	calibrate: bool, // whether calibration is enabled.
-	inner: AtomicUsize,
-	duration: u16,
+	inner: AtomicU64,
+	/// Planned durations of steps.
+	durations: Vec<(TransitionStep, TransitionTimestamp, StepDuration)>,
 }
 
 impl Step {
-	fn load(&self) -> u64 { self.inner.load(AtomicOrdering::SeqCst) as u64 }
+	fn load(&self) -> u64 { self.inner.load(AtomicOrdering::SeqCst) }
+
+	/// Finds the remaining duration of the current step. Panics if there was a counter under- or
+	/// overflow.
 	fn duration_remaining(&self) -> Duration {
-		let now = unix_now();
-		let expected_seconds = self.load()
-			.checked_add(1)
-			.and_then(|ctr| ctr.checked_mul(self.duration as u64))
-			.map(Duration::from_secs);
-
-		match expected_seconds {
-			Some(step_end) if step_end > now => step_end - now,
-			Some(_) => Duration::from_secs(0),
-			None => {
-				let ctr = self.load();
-				error!(target: "engine", "Step counter is too high: {}, aborting", ctr);
-				panic!("step counter is too high: {}", ctr)
-			},
-		}
+		self.opt_duration_remaining().unwrap_or_else(|| {
+			let ctr = self.load();
+			error!(target: "engine", "Step counter under- or overflow: {}, aborting", ctr);
+			panic!("step counter under- or overflow: {}", ctr)
+		})
+	}
 
+	/// Finds the remaining duration of the current step. Returns `None` if there was a counter
+	/// under- or overflow.
+	fn opt_duration_remaining(&self) -> Option<Duration> {
+		let next_step = self.load().checked_add(1)?;
+		let (mut prev_dur, mut prev_step, mut prev_time) =
+			self.durations.first().expect("durations cannot be empty");
+		for (step, time, dur) in self.durations.iter().skip(1) {
+ 			if *step >= next_step {
+				break;
+			}
+			prev_step = *step;
+			prev_time = *time;
+			prev_dur = *dur;
+		}
```

Comment:
Maybe simpler:
```rust
let (prev_dur, prev_step, prev_time) = self.durations.iter()
    .take_while(|(_, step, _)| step < next_step)
    .last()
    .expect("durations cannot be empty")
    .clone();
```
(Also in `opt_calibrate`.)

Reply:
Actually, it's `(step, _, _)`, which brings the question: do you think there should be a struct instead of the triple?

Reply:
Ah, sorry! But then it should also be:
```rust
let (prev_step, prev_dur, prev_time) = …
```
A struct would avoid that kind of confusion, yes, but it's also a bit more verbose. Your call.

Reply:
It's `(step, time, dur)` and I dropped `prev_` as redundant.